Add support for instantiating a new torch.dtype in python

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116295
* __->__ #116294

Summary:
This allows us to add new custom dtypes in the python frontend such as torch.int4,
it will be used in https://github.com/pytorch-labs/ao/pull/13

Test Plan:
python test/test_quantization.py -k test_new_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Add support for instantiating a new torch.dtype in python

Summary:
This allows us to add new custom dtypes in the python frontend such as torch.int4,
it will be used in https://github.com/pytorch-labs/ao/pull/13

Test Plan:
python test/test_quantization.py -k test_new_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for instantiating a new torch.dtype in python"


Summary:
This allows us to add new custom dtypes in the python frontend such as torch.int4,
it will be used in https://github.com/pytorch-labs/ao/pull/13

Test Plan:
python test/test_quantization.py -k test_new_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vadimkantorov commented:
Could this be used to instantiate tiled dtypes? as proposed in
- https://github.com/pytorch/pytorch/issues/104702

jerryzh168 commented:
> Could this be used to instantiate tiled dtypes? as proposed in
> 
> * [[feature request] Specialized memory layouts and wide blocked/tiled dtypes for cublasLt/onednn: e.g. torch.float16x32 / torch.int8x32 / torch.bits1x512 (akin to torch.quint2x4) #104702](https://github.com/pytorch/pytorch/issues/104702)

I'm not sure if tiling is a part of dtype? but if it is, we can use this to instantiate it, although this is just a simple frontend API that doesn't not have much functionalities.

soulitzer commented:
It may be confusing to offer this API since newly created dtype does not behave as expected to general users: when translated to cpp it becomes the original scalar_type and would no longer print out what the user specifies. 

A private API would be better... but not sure that is the best either - even if your subclass overrides .dtype in python so that it would call this private API there, it may be confusing to have that return a dtype in which the scalar_type and str do not match. It would be confusing to the user what they are allowed to do with that dtype.

jerryzh168 commented:
> It may be confusing to offer this API this way since newly created dtype does not behave as expected to general users: when translated to cpp it becomes the original scalar_type and would no longer print out what the user specifies.
> 

I feel this might be OK, and that is what we want as well, the new dtypes are front end only, when translated to cpp, we want some pre-existing dtypes like uint8, bits8 etc.

soulitzer commented:
Having int4 internally represented as int8 makes sense to me. I think the issue is that in the current state, it is only possible to translate the frontend-dtype int4 to int8, but the reverse is not possible. Once it becomes a scalar_type int8 it loses the information about being int4, so that if that when surfaced to the user later, it would no longer be int4, and that may be confusing.

jerryzh168 commented:
> Having int4 internally represented as int8 makes sense to me. I think the issue is that in the current state, it is only possible to translate the frontend-dtype int4 to int8, but the reverse is not possible. Once it becomes a scalar_type int8 it loses the information about being int4, so that if that when surfaced to the user later, it would no longer be int4, and that may be confusing.

what is the use case for translating int8 back to int4? the use case I'm thinking of is that we have frontend int4 dtype tensor and when we use it in cpp or when we lower it to some backend, we'll store some information in operators, e.g. int4_op(int8_tensor, ...), that's how we recognize the int4 dtype tensor when it is desugared.

soulitzer commented:
If user creates `torch.dtype(torch.int8, "int4")` and passes to a factory function, that would create an plain int8 tensor. 
I'm not saying we necessarily want to support this case, its just that users will inevitably try to do this with the ctor and it would end up doing something unexpected.

jerryzh168 commented:
> If user creates `torch.dtype(torch.int8, "int4")` and passes to a factory function, that would create an plain int8 tensor. I'm not saying we necessarily want to support this case, its just that users will inevitably try to do this with the ctor and it would end up doing something unexpected.

I see, yeah we want to error out in these unsupported cases

jerryzh168 committed:
Update on "Add support for instantiating a new torch.dtype in python"


Summary:
This allows us to add new custom dtypes in the python frontend such as torch.int4,
it will be used in https://github.com/pytorch-labs/ao/pull/13

Test Plan:
python test/test_quantization.py -k test_new_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "Add support for instantiating a new torch.dtype in python"


Summary:
This allows us to add new custom dtypes in the python frontend such as torch.int4,
it will be used in https://github.com/pytorch-labs/ao/pull/13

Test Plan:
python test/test_quantization.py -k test_new_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

soulitzer commented on torch/csrc/Dtype.cpp:
```diff
@@ -157,6 +173,16 @@ PyTypeObject THPDtypeType = {
     nullptr, /* tp_new */
 };
 
+static PyMethodDef TorchMethods[] = { // NOLINT
```
This isn't used anywhere right?

soulitzer commented on torch/csrc/Dtype.cpp:
```diff
@@ -25,6 +28,19 @@ PyObject* THPDtype_New(at::ScalarType scalar_type, const std::string& name) {
   END_HANDLE_TH_ERRORS
 }
 
+PyObject* THPDtype_pynew(PyObject* type, PyObject* args, PyObject* kwargs) {
```
nit: Maybe rename to something like THPDtype_pyNewCustomDtype 
THPDtype_pynew reads like something that goes in tp_new

A small comment mentioning the limitations of this API would be good.

jerryzh168 committed:
Update on "Add support for instantiating a new torch.dtype in python"


Summary:
This allows us to add new custom dtypes in the python frontend such as torch.int4,
it will be used in https://github.com/pytorch-labs/ao/pull/13

Test Plan:
python test/test_quantization.py -k test_new_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vadimkantorov commented:
> I'm not sure if tiling is a part of dtype

I think in NumPy it is indeed a part of dtype https://numpy.org/doc/stable/reference/arrays.dtypes.html, e.g. for fixed-width strings (and also some thing like sub-array? and also in a way for custom struct-like dtypes which are reinterpretation of a fixed-width blob) @ezyang

ezyang reviewed (changes requested):
I'm sorry, can we have a more detailed spec for this?

ezyang commented:
I don't think this does what you want.

What this does is allow you to generate a fresh Python binding to a pre-existing PyTorch dtype. But there isn't any evidence in this PR that (1) this works correctly (THPDtype_New was always written under the assumption that we call it once at static init time per ScalarType) and (2) that it is sufficient to solve whatever use case you want.

Instead of stacking diffs, I think to start I want to see some end-to-end description of how exactly this is going to use. For example, how are you going to do dispatch on this? Are you allowed to pass these to stock constructors? What happens if you query the dtype of a tensor that was created this way? A prototype may be necessary but I don't think it's even clear how this is supposed to work, even in principle.

jerryzh168 commented:
> I don't think this does what you want.
> 
> What this does is allow you to generate a fresh Python binding to a pre-existing PyTorch dtype. But there isn't any evidence in this PR that (1) this works correctly (THPDtype_New was always written under the assumption that we call it once at static init time per ScalarType) and (2) that it is sufficient to solve whatever use case you want.
> 
> Instead of stacking diffs, I think to start I want to see some end-to-end description of how exactly this is going to use. For example, how are you going to do dispatch on this? Are you allowed to pass these to stock constructors? What happens if you query the dtype of a tensor that was created this way? A prototype may be necessary but I don't think it's even clear how this is supposed to work, even in principle.

oh yeah we do have a prototype here: https://github.com/pytorch-labs/ao/pull/13 (see tests), there is no guarantee for any of these things to work, the idea is that the features will be added on demand in the tensor subclass (or factory functions etc. - need to decide later). The main goal for the PR is just to allow us to create a custom dtype so that it works for some of the quantization flows like GPU quantization with torch.compile lowering

jerryzh168 commented:
> (THPDtype_New was always written under the assumption that we call it once at static init time per ScalarType)

what would happen if we call this multiple times?

ezyang commented:
The prototype doesn't say anything about how this design for dtype UX will work. Are you trying to solve the .to(...) problem on the Uint4Tensor subclass? Or something else?

jerryzh168 commented:
> The prototype doesn't say anything about how this design for dtype UX will work. Are you trying to solve the .to(...) problem on the Uint4Tensor subclass? Or something else?

right I'm planning to figure these out a bit later, only thing that this enabled right now is:
```
        return torch.Tensor._make_wrapper_subclass(cls, up_size(elem.shape), dtype=torch.uint4, **kwargs)
```

and then we can go through torch.compile I think

jerryzh168 commented:
not going to do this since we added uint1 to uint7 native dtypes https://github.com/pytorch/pytorch/pull/117208
